server/src/utils.ts

```diff
@@ -78,3 +204,93 @@ function generateDiagnostic(severity: DiagnosticSeverity,
   }
   return diagnostic
 }
+
+function calculateElementPositions(element: any): Position[] {
+  // Calculate positions accounting for the zero-based convention used by
+  // vscode
+  const startPosition: Position = {
+    line: element.lineNumber - 1,
+    character: element.columnNumber - 1
+  }
+  const endPosition: Position = {
+    line: element.nextSibling.lineNumber - 1,
+    character: element.nextSibling.columnNumber - 1
+  }
+
+  return [startPosition, endPosition]
+}
+
+export interface ValidationRequest {
+  textDocument: TextDocument
+  version: number
+}
+
+export class ValidationQueue {
+  private queue: ValidationRequest[]
+  private timer: NodeJS.Immediate | undefined
+
+  constructor(private readonly connection: Connection) {
+    this.queue = []
+  }
+
+  public addRequest(request: ValidationRequest): void {
+    this.dropOldVersions(request)
+    this.queue.push(request)
+    this.trigger()
+  }
+
+  private dropOldVersions(request: ValidationRequest): void {
+    // It's possible to get validation requests for the same document before
+    // we've processed older ones. We can use the document version (which
+    // increases after each change, even if it's undo / redo) to prune the queue
+    const updatedQueue = this.queue.filter(entry => {
+      const isOlderVersion = (entry.textDocument.uri === request.textDocument.uri) && (entry.version < request.version)
+      const isDifferentDocument = (entry.textDocument.uri !== request.textDocument.uri)
+      return (!isOlderVersion || isDifferentDocument)
+    })
+
+    this.queue = updatedQueue
+  }
+
+  private trigger(): void {
+    if (this.timer !== undefined || this.queue.length === 0) {
+      // Either the queue is empty, or we're already set to process the next
+      // entry
+      return
+    }
+
+    this.timer = setImmediate(() => {
+      this.processQueue().finally(() => {
+        this.timer = undefined
+        this.trigger()
+      })
+    })
+  }
+
+  private async processQueue(): Promise<void> {
+    const request = this.queue.shift()
+    if (request == null) {
+      return
+    }
+    const textDocument = request.textDocument
+    let workspaceFolders = await this.connection.workspace.getWorkspaceFolders()
+    if (workspaceFolders == null) {
+      workspaceFolders = []
+    }
+    const diagnostics: Diagnostic[] = []
+    const xmlData = parseXMLString(textDocument)
+    const knownModules = await getCurrentModules(workspaceFolders)
```

Comment:
🤔  - I think it would require implementing a specific [request](https://microsoft.github.io/language-server-protocol/specification#requestMessage) / response exchange between the two. I haven't really looked at the specifics of using that mechanism, though. But, in general, I do think it's possible based on the protocol spec.

Reply:
Bummer that it'd require a bit of a dance!